src/url.cpp

```diff
@@ -6,6 +6,93 @@
 #include <string>
 
 namespace ada {
+  ada_really_inline bool url::parse_path(std::string_view input) {
+    std::string tmp_buffer;
+    std::string_view internal_input;
+    if(std::any_of(input.begin(), input.end(), ada::unicode::is_ascii_tab_or_newline)) {
+      tmp_buffer = input;
+      // Optimization opportunity: Instead of copying and then pruning, we could just directly
+      // build the string from user_input.
+      helpers::remove_ascii_tab_or_newline(tmp_buffer);
+      internal_input = tmp_buffer;
+    } else {
+      internal_input = input;
+    }
+
+    // If url is special, then:
+    if (is_special()) {
+      if(internal_input.empty()) {
+        path = "/";
+      } else if((internal_input[0] == '/') ||(internal_input[0] == '\\')){
+        parse_prepared_path(internal_input.substr(1));
+      } else {
+        parse_prepared_path(internal_input);
+      }
+    } else if (!internal_input.empty()) {
+      if(internal_input[0] == '/') {
+        parse_prepared_path(internal_input.substr(1));
+      } else {
+        parse_prepared_path(internal_input);
+      }
+    } else {
+      if(!host.has_value()) {
+        path = "/";
+      }
+    }
+    return true;
+  }
+
+  ada_really_inline bool url::parse_prepared_path(std::string_view input) {
+    path="";
+    bool needs_percent_encoding = input.end() != std::find_if(input.begin(), input.end(),
+    [](uint8_t c) {
+      return c <= 35 || c > 122 || c == 96 || (c>59 && c<=63) ; } );
```

Comment:
Can you add a comment for this? It’s hard to keep track

Reply:
I did add a comment, yes. It is a heuristic to quickly determine if we need percent encoding (we often do not in practice so it is worth checking).